core/deadline.go

```diff
@@ -34,18 +36,135 @@ type slotTimeProvider interface {
+	// Add adds a duty to be notified of the Deadline via C.
+	// Note that duties will be deduplicated and only a single duty will be provided via C.
 	Add(duty Duty)
 
 	// C returns the same read channel every time and contains deadlined duties.
 	// It should only be called by a single goroutine.
 	C() <-chan Duty
 }
 
+// Deadline implements the Deadliner interface.
+type Deadline struct {
+	mu           sync.Mutex
+	genesisTime  time.Time
+	slotDuration time.Duration
+
+	dutyChan     chan Duty
+	deadlineChan chan Duty
+
+	// duties represents a set since duty deduplication is required.
+	duties map[Duty]bool
+}
+
+// NewDeadliner returns a new instance of Deadline.
+// It runs a goroutine which is responsible for reading and storing duties,
+// and sending the deadlined duty to receiver's deadlineChan.
+func NewDeadliner(ctx context.Context, eth2Cl slotTimeProvider) (*Deadline, error) {
+	genesis, err := eth2Cl.GenesisTime(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	duration, err := eth2Cl.SlotDuration(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	d := &Deadline{
+		genesisTime:  genesis,
+		slotDuration: duration,
+		dutyChan:     make(chan Duty),
+		deadlineChan: make(chan Duty),
+		duties:       make(map[Duty]bool),
+	}
+
+	go func() {
+		var (
+			deadlineTimer <-chan time.Time
+			minDuty       Duty
+		)
+
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case duty := <-d.dutyChan:
+				if minDuty.Type == DutyUnknown {
+					// Initialise minDuty and deadlineTimer
+					minDuty = duty
+					deadlineTimer = time.After(time.Until(d.deadlineTime(minDuty)))
+				}
+				d.duties[duty] = true
+			case <-deadlineTimer:
+				// Send deadlined duty to receiver
+				d.deadlineChan <- minDuty
+
+				// Delete duty whose deadline has passed
+				delete(d.duties, minDuty)
+
+				// New min duty for next deadline
+				md, ok := d.getMinDuty()
+				if !ok {
+					continue
+				}
+
+				minDuty = md
+				deadlineTimer = time.After(time.Until(d.deadlineTime(minDuty)))
+			}
+		}
+	}()
+
+	return d, nil
+}
+
+// Add adds a duty to be notified of the deadline.
+func (d *Deadline) Add(duty Duty) {
+	d.dutyChan <- duty
+}
+
+// C returns the deadline channel.
+func (d *Deadline) C() <-chan Duty {
+	return d.deadlineChan
+}
+
+// getMinDuty gets the duty to process next.
+// It selects duty with minimum slot. If slots are equal, then it selects the duty with minimum DutyType.
+func (d *Deadline) getMinDuty() (Duty, bool) {
+	minDuty := Duty{Slot: math.MaxInt64, Type: dutySentinel}
+	for duty := range d.duties {
+		if duty.Slot < minDuty.Slot || ((duty.Slot == minDuty.Slot) && (duty.Type < minDuty.Type)) {
+			minDuty = duty
+		}
+	}
+
+	if minDuty.Slot == math.MaxInt64 {
+		return minDuty, false
+	}
+
+	return minDuty, true
+}
+
+// deadlineTime returns the deadline time of the duty.
+func (d *Deadline) deadlineTime(duty Duty) time.Time {
```

Comment:
do not duplicate this logic, suggest passing in `deadlineFunc` as argument to constructor.

Reply:
sure